File: fitness_calc.py
"""
Fitness calculation and data extraction module for EGAGP.
Optimized using NumPy for efficient array operations.
"""

import numpy as np
from typing import Tuple


class FitnessCalc:
    """Handles data extraction and storage for the genetic algorithm."""

    # Component communication cost matrix
    C: np.ndarray = None
    # Weight of each component
    W: np.ndarray = None
    # Machine communication cost matrix
    B: np.ndarray = None
    # Machine capacity array
    M: np.ndarray = None

    number_of_machines: int = 0
    number_of_vertices: int = 0

    @classmethod
    def extract_data(cls, path: str) -> None:
        """
        Extract graph and machine data from file.

        Args:
            path: Path to the data file
        """
        with open(path, 'r') as file:
            lines = [line.strip() for line in file.readlines()]

        idx = 0

        # Read graph (components) data
        n = int(lines[idx])
        idx += 1
        m = int(lines[idx])
        idx += 1

        cls.number_of_vertices = n
        cls.C = np.zeros((n, n), dtype=np.float64)

        # Read edges
        for _ in range(m):
            parts = lines[idx].split()
            x, y, cost = int(parts[0]), int(parts[1]), float(parts[2])
            cls.C[x, y] = cost
            cls.C[y, x] = cost  # Symmetric
            idx += 1

        # Read component weights
        cls.W = np.array([float(x) for x in lines[idx].split()], dtype=np.float64)
        idx += 1

        # Read machine data
        n = int(lines[idx])
        idx += 1
        m = int(lines[idx])
        idx += 1

        cls.number_of_machines = n
        cls.B = np.full((n, n), 1000000.0, dtype=np.float64)

        # Read machine connections
        for _ in range(m):
            parts = lines[idx].split()
            x, y, cost = int(parts[0]), int(parts[1]), float(parts[2])
            cls.B[x, y] = cost
            cls.B[y, x] = cost  # Symmetric
            idx += 1

        # Read machine capacities
        cls.M = np.array([float(x) for x in lines[idx].split()], dtype=np.float64)

File: java/FitnessCalc.java
import java.io.BufferedReader;
import java.io.FileReader;
 
public class FitnessCalc {
    public static double[][] C;       // Component communication cost
    public static double[] W;         // Weight of component
    public static double[][] B;       // Machine communication cost
    public static double[] M;         // Machine capacity

    public static int numberOfMachines, numberOfVertex;

    public static void extractData(String path) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(path));
            String line;
            int n, m;
            int x, y;
            String[] array;

            //1st Graph
            n = Integer.parseInt(br.readLine());
            m = Integer.parseInt(br.readLine());
            numberOfVertex = n;
            C = new double[n][n];
            W = new double[n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    C[i][j] = 0;
                }
            }
            for (int i = 0; i < m; i++) {
                line = br.readLine();
                array = line.split(" ");
                x = Integer.parseInt(array[0]);
                y = Integer.parseInt(array[1]);
                C[x][y] = Double.parseDouble(array[2]);
                C[y][x] = Double.parseDouble(array[2]);
            }
            line = br.readLine();
            array = line.split(" ");
            for (int i = 0; i < array.length; i++) {
                W[i] = Double.parseDouble(array[i]);
            }

            n = Integer.parseInt(br.readLine());
            m = Integer.parseInt(br.readLine());
            numberOfMachines = n;
            B = new double[n][n];
            M = new double[n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    B[i][j] = 1000000.0;
                }
            }
            for (int i = 0; i < m; i++) {
                line = br.readLine();
                array = line.split(" ");
                x = Integer.parseInt(array[0]);
                y = Integer.parseInt(array[1]);
                B[x][y] = Double.parseDouble(array[2]);
                B[y][x] = Double.parseDouble(array[2]);
            }
            line = br.readLine();
            array = line.split(" ");
            for (int i = 0; i < array.length; i++) {
                M[i] = Double.parseDouble(array[i]);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}